### ISSUE TITLE
Question regarding hashing

### ISSUE THREAD

**Issue by emaildano** [gatsbyjs/gatsby#23988]
## Summary

I have a question regarding hashing and the use of md5. Correct me if I'm wrong but it looks like Gatsby uses md5 as the algorithm of choice for naming files and folders in the local cache. I'm not too familiar with Gatsby "core" but reviewing the bootstrap files and searching the source it looks like this is the only used aside from sha256 for telemetry data.

Why md5?

Hashing is a crucial part of Gatsby's build and develop process but I question the use of md5 for a few reasons. Again, maybe there's a very good reason I'm not aware of so feel free to tell me I'm spinning my wheels.

The md5 algorithm is a cryptographic hash suitable for non-cryptographic purposes and this seems to be the use case. Since there are faster algorithms available, why not use those?

Sometimes we're talking milliseconds difference but it adds up over all the various builds.

I wonder if it's possible to use a different hash such as Adler32 or crc32b which are designed for this purpose and operate much faster. Beyond that there are even faster options which are not supported on most systems by default so they would have to be installed as a dependency.

One of the options that comes to mind is [xxHash](https://github.com/Cyan4973/xxHash).

xxHash is crazy fast compared to md5 based on their [benchmarks](https://github.com/Cyan4973/xxHash#benchmarks).

## Recap

In conclusion 🤓, maybe the hash choice isn't a big deal or a major factor in performance. But maybe... it is? I ask because it could improve build and development time.

**Comment by emaildano** [comment]
Reference:

> On 64-bit systems, MD5 will compute at about the same speed (multi-threaded programs aside) of 0.33Gb\s but the 64-bit implementation of xxHash will compute at about 13Gb\s (780Gb p\minute)! That’s about 30 times faster than MD5 and 2.5 times faster than the already very fast 32-bit xxHash algorithm!

https://www.quickhash-gui.org/what-is-the-newly-added-xxhash-algorithm/

**Comment by pieh** [comment]
I'm not sure if using `md5` was very intentional for those cases (other than being very well known and established).

Do you know what hashing algo is used by `webpack` to create hashes for assets it emits? Our usecase is very similar, because we don't really use it for cryptographic purposes but mostly as a way to identify some changes (used when we compare hashes) or to "communicate" content changes to hosting/browsers (so we can use permanent caching server rules)

**Comment by pieh** [comment]
Things for us to consider I guess are collision characteristics and length of generated hashes for the algos you mentioned

**Comment by pieh** [comment]
For the performance part - I don't think I ever looked specifically into that, but worth doing profiling build ( https://www.gatsbyjs.org/docs/debugging-the-build-process/#chrome-devtools-for-node ) and record cpu profile and look up how much of the build is spend on generating those hashes (so look for `crypto` calls there). It will not be 100% accurate because just attaching debugger does change things, but is good enough approximation

**Comment by emaildano** [comment]
@pieh Thanks for benchmarking info, that sounds like it would be helpful for testing this idea. I also wasn't aware that the hashing choice was up to webpack so I'll look into that too.

Ps: regarding hash length that's less important as long as it passes that hashing test the hashing community recommends.

Thanks! 🍻
